common/lib/xmodule/xmodule/x_module.py

```diff
@@ -1859,8 +1861,14 @@ def __init__(
         self.descriptor_runtime = descriptor_runtime
         self.rebind_noauth_module_to_user = rebind_noauth_module_to_user
 
-        if user:
-            self.user_id = user.id
+        # Fetch user data from the user service
+        user_service = self.service(self, 'user')
```

Comment:
Is it ok to do this here?

I tried creating these as on-demand `@property` attributes, but that didn't work because of how CombinedSystem delegates attribute fetching between the ModuleSystem and DescriptorSystem.

Reply:
> Here for the shim code they can be accessed via self._services.

Excellent, thanks for this info! Glad to know I was almost on track here.

> Hmm, can you share the exact code and error? I do see that ajax_url below is a property.

I'm not sure what was causing an issue previously -- I wasn't able to reproduce it when I used your code above. So all good to use `@property` values here.

However, I did encounter a problem when replacing `user_id` attribute with a property, see https://github.com/edx/XBlock/pull/542